File: src/main/java/frc/robot/subsystems/shooter/ShooterIO.java
package frc.robot.subsystems.shooter;

import static edu.wpi.first.units.Units.RotationsPerSecond;

import org.littletonrobotics.junction.AutoLog;

import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.units.measure.AngularVelocity;

public interface ShooterIO {
    @AutoLog
    public static class ShooterIOInputs {
        public AngularVelocity flywheelVelocity = RotationsPerSecond.of(0);
        public double[] flywheelAppliedVoltages = new double[3];
        public double[] flywheelCurrents = new double[3];
        public double[] flywheelTemperatures = new double[3];

        public Rotation2d hoodPosition = Rotation2d.kZero;
        public double hoodAppliedVoltage = 0.0;
        public double hoodCurrent = 0.0;
        public double hoodTemperature = 0.0;
        public AngularVelocity hoodVelocity = RotationsPerSecond.of(0);
    }

    public default void setFlywheelVoltage(double voltage) {}

    public default void setHoodVoltage(double voltage) {}

    public default void setFlywheelVelocity(AngularVelocity velocity) {}

    public default void setHoodPosition(Rotation2d position) {}

    public default void updateInputs(ShooterIOInputs inputs) {}
}

File: src/main/java/frc/robot/subsystems/shooter/ShooterIO.kt
package frc.robot.subsystems.shooter

import edu.wpi.first.units.Units.RotationsPerSecond
import edu.wpi.first.units.measure.AngularVelocity
import edu.wpi.first.math.geometry.Rotation2d
import com.lobstahbots.junction.AutoLogKt
import com.lobstahbots.units.*

interface ShooterIO {
    @AutoLogKt
    open class ShooterIOInputs {
        var flywheelVelocity: AngularVelocity = 0.0.rpm
        var flywheelSetpoint: AngularVelocity = 0.0.rpm
        var flywheelAppliedVoltages = doubleArrayOf(0.0, 0.0, 0.0)
        var flywheelCurrents = doubleArrayOf(0.0, 0.0, 0.0)
        var flywheelTemperatures = doubleArrayOf(0.0, 0.0, 0.0)

        var hoodPosition: Rotation2d = Rotation2d.kZero
        var encoderPosition: Rotation2d = Rotation2d.kZero
        var hoodAppliedVoltage = 0.0
        var hoodCurrent = 0.0
        var hoodTemperature = 0.0
        var hoodVelocity: AngularVelocity = 0.0.rpm

        var useProfile = true
    }

    fun setFlywheelVoltage(voltage: Double) {}

    fun setHoodVoltage(voltage: Double) {}

    fun setFlywheelVelocity(velocity: AngularVelocity) {}

    fun setHoodPosition(position: Rotation2d) {}

    fun updateInputs(inputs: ShooterIOInputs) {}

    fun resetEncoder(position: Rotation2d) {}
}